nosleep.1
.TH NOSLEEP 1 "Sep 20, 2011" "0.1" "CYGWIN"


.SH NAME
nosleep - Run a program while inhibiting computer sleep or hibernation


.SH SYNOPSIS
.B nosleep
[OPTIONS]
.I command
.RI [ args ]


.SH DESCRIPTION
.B nosleep
runs the supplied command,
.I command 
.RI [ args ],
while inhibiting the computer from sleeping or hibernating until the command
finishes executing.

.B nosleep
does not guarantee that the computer won't go to sleep.  By default, it just
prevents the computer from automatically putting itself to sleep during periods
of no user input (although it still allows the display to turn off).  If the
user closes the laptop lid or presses the power button, the computer will still
behave as its power management plan dictates--typically by sleeping,
hibernating, or shutting down.

Inhibiting idle sleep is a safe mode of operation, since it won't prevent the
computer from shutting down if needs to to conserve power, and for many
applications this mode is sufficient.  When it's not, options are available,
below, to more aggressively inhibit sleep or hibernation.  But before you use
any of them, you must read the:


.SH WARNINGS AND DISCLAIMER
Aggressively inhibiting sleep or hibernation carries at least the following
disadvantages and risks:

.TP
1.
If the host is running on battery, preventing sleep, hibernation, or display
shutoff could cause the host to run out of power and shut down uncleanly,
potentially causing data loss.

.TP
2.
Users may become confused or frustrated if the usual sleep behavior that they
expect, such as going to sleep when the laptop lid is closed, stops working or
only works some of the time.

.TP
3.
.B nosleep
still can't guarantee that the
.I command
will complete, since it can't prevent users or other processes from shutting
down the host.  For example, a user who tries and fails to put the computer to
sleep may decide to shut it down instead.

.LP
The author disclaims all responsibility for any losses you may incur by using
the more aggressive options below.


.SH OPTIONS
.TP
--awaymode
Make the computer go into "away mode" when it would normally go to sleep.
The computer will appear to be asleep, but the
.I command
will continue executing.  The Windows documentation[1] says: "Applications that
do not require critical background processing or that run on portable computers
should not enable away mode because it prevents the system from conserving power
by entering true sleep."  Use your power for good, not evil.

Away mode is only available if it's allowed by the active power plan, and there
is no direct way for the user to determine if this is the case.  For more
information see the section "Away Mode and Windows Vista Power Policy" in [2].

.I Windows Server 2003 and Windows XP/2000:
Away mode is not supported.  This option will have no effect.

.TP
--display
Prevent the display from dimming or turning off during idle periods.

.TP
--ifacpower
Options that follow this one will only take effect if the host is running on AC
power. The AC power state is only checked when
.B nosleep
starts; if AC power is removed after that,
.B nosleep
won't detect it.

.SH EXAMPLES
.IP
nosleep rsync dir1 dir2
.LP
will run rsync dir1 dir2, preventing idle sleep until rsync exits.
.IP
nosleep --ifacpower --display rsync dir1 dir2
.LP
will prevent idle sleep as before, and will also keep the display from turning
off if the computer is running on AC power when
.B nosleep
starts.

.SH RETURN VALUE
.B nosleep
returns the exit value of
.IR command ,
or 128 if it encounters a fatal error.

.SH BUGS
Please send bug reports to the Cygwin mailing list <cygwin@cygwin.com>.

.SH REFERENCES
.TP
[1]
http://msdn.microsoft.com/en-us/library/aa373208(VS.85).aspx
.TP
[2]
http://download.microsoft.com/download/9/c/5/9c5b2167-8017-4bae-9fde-d599bac8184a/AwayMode.doc
.SH AUTHOR
Andrew E. Schulman <andrex.e.schulman@gmail.com>


ChangeLog
Change log file for nosleep

2011-11-12  Version 0.1.5
* Fix Possible crash in is_AC_power_on(). (#893960; by Christian Franke)

2011-09-30  Version 0.1.4
* Warn but continue executing if sleep inhibition fails.
* Add first functional test.

2011-09-24  Version 0.1.3
* Remove all implementation of --powerplan option.  That option is targeted
  for the 0.2 series.

2011-09-24  Version 0.1.2
* Fix reordering of command line arguments.
* Add basic make install target.

2011-09-22  Version 0.1.1
* Add README and ChangeLog.

2011-09-22  Version 0.1
* Initial release:
  * Cygwin only.
  * All options implemented except --powerplan.



nosleep.c
/**
 * nosleep.c
 */

/*
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
/* Copyright 2011 Andrew E. Schulman */

#include <argp.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <unistd.h>
#include "nosleep.h"

/* Program docs */
const char *argp_program_version = "nosleep v0.1.5 | Copyright 2011 Andrew E. Schulman";
const char *argp_program_bug_address = "the Cygwin mailing list <cygwin@cygwin.com>";
static char doc[]         = "Run a command while inhibiting computer sleep or hibernation.";
static char args_doc[]    = "command [args]"; /* non-option arguments */

/* Command-line options */
static struct argp_option options[] = {
  { "awaymode",  'a', 0, 0, "Force away mode instead of sleep mode" },
  { "display",   'd', 0, 0, "Keep the display on" },
  { "ifacpower", 'i', 0, 0, "Following options only apply if AC power is on" },
  { 0 }
};

/* argp parser */
static error_t parse_opt(int, char *, struct argp_state *);
static struct argp argp = {options, parse_opt, args_doc, doc};

/* Option states, as determined on the command line */
#define NOSLEEP_OPTION_OFF     0
#define NOSLEEP_OPTION_ON      1
#define NOSLEEP_OPTION_AC_ONLY 2

#define NOSLEEP_EXIT_ERROR 128
#define ERROR_LEN          256


/**
 * main()
 */

int
main(int argc, char *argv[]) {

  char error[ERROR_LEN] = "unknown error";
  struct power_options poweropts;
  int firstarg, child_status, i;

  /* Process command-line options */
  poweropts.noidlesleep = NOSLEEP_OPTION_ON;
  poweropts.awaymode    = NOSLEEP_OPTION_OFF;
  poweropts.display     = NOSLEEP_OPTION_OFF;

  if (argp_parse(&argp, argc, argv, ARGP_IN_ORDER, &firstarg, &poweropts))
    exit(NOSLEEP_EXIT_ERROR);

  /* Filter out AC-only options, if AC power isn't on */
  if (!is_AC_power_on()) {
    poweropts.noidlesleep &= ~NOSLEEP_OPTION_AC_ONLY;
    poweropts.awaymode    &= ~NOSLEEP_OPTION_AC_ONLY;
    poweropts.display     &= ~NOSLEEP_OPTION_AC_ONLY;
  }

  /* Inhibit sleep/hibernation */
  switch (set_sleep_inhibition_state(poweropts, error, sizeof(error))) {
  case NOSLEEP_OP_WARNING:
    fprintf(stderr, "%s: Warning: %s\n", argv[0], error);
    break;
  case NOSLEEP_OP_ERROR:
    fprintf(stderr, "%s: Error: %s\n", argv[0], error);
    exit(NOSLEEP_EXIT_ERROR);
    break;
  }

  /* Start child process */
  switch (fork()) {
  case 0: /* child */
    execvp(argv[firstarg], argv+firstarg);
    fprintf(stderr, "%s: Error: failed to execute command:", argv[0]);
    for (i = firstarg; i<argc; ++i)
      fprintf(stderr, " %s", argv[i]);
    fprintf(stderr, "\n");
    exit(NOSLEEP_EXIT_ERROR);
    break;
  case -1: /* parent (error) */
    fprintf(stderr, "%s: Error: could not fork\n", argv[0]);
    exit(NOSLEEP_EXIT_ERROR);
    break;
  }

  /* Wait for child to exit */
  wait(&child_status);

  /* Restore power options */
  switch(clear_sleep_inhibition_state(poweropts, error, sizeof(error))) {
  case NOSLEEP_OP_WARNING:
    fprintf(stderr, "%s: Warning: %s\n", argv[0], error);
    break;
  case NOSLEEP_OP_ERROR:
    fprintf(stderr, "%s: Error: %s\n", argv[0], error);
    exit(NOSLEEP_EXIT_ERROR);
    break;
  }

  /* Exit */
  if (WIFEXITED(child_status))
    exit(WEXITSTATUS(child_status));
  else
    exit(NOSLEEP_EXIT_ERROR);
}


/*
 * parse_opt():  Parse a single option.
 */

static error_t
parse_opt(int key, char *arg, struct argp_state *state) {

  static int ac_state = NOSLEEP_OPTION_ON;
  struct power_options *poweropts = state->input;

  switch (key) {
  case 'a':
    poweropts->awaymode  = ac_state;
    break;
  case 'd':
    poweropts->display   = ac_state;
    break;
  case 'i':
    ac_state = NOSLEEP_OPTION_AC_ONLY;
    break;

  case ARGP_KEY_ARG:
  case ARGP_KEY_ARGS:
    /* don't handle non-option arguments here; send them back to main() */
    return ARGP_ERR_UNKNOWN;
    break;

  case ARGP_KEY_NO_ARGS:
    argp_usage(state);
    break;

  default:
    return ARGP_ERR_UNKNOWN;
  }

  return 0;
}
